src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateSignature.cs

```diff
@@ -0,0 +1,314 @@
+                                jwtToken.EncodedToken,
+                                JwtTokenUtilities.SafeLogJwtToken)
+                            ),
+                        typeof(SecurityTokenInvalidSignatureException),
+                        new StackFrame()));
+
+            SecurityKey? key = null;
+            if (validationParameters.IssuerSigningKeyResolver is not null)
+            {
+                key = validationParameters.IssuerSigningKeyResolver(
+                    jwtToken.EncodedToken,
+                    jwtToken,
+                    jwtToken.Kid,
+                    validationParameters,
+                    configuration,
+                    callContext);
+            }
+            else
+            {
+                // Resolve the key using the token's 'kid' and 'x5t' headers.
+                // Fall back to the validation parameters' keys if configuration keys are not set.
+                key = JwtTokenUtilities.ResolveTokenSigningKey(jwtToken.Kid, jwtToken.X5t, configuration?.SigningKeys)
+                    ?? JwtTokenUtilities.ResolveTokenSigningKey(jwtToken.Kid, jwtToken.X5t, validationParameters.IssuerSigningKeys);
+            }
+
+            if (key is not null)
+            {
+                jwtToken.SigningKey = key;
+
+                // If the key is found, validate the signature.
+                return ValidateSignatureWithKey(jwtToken, key, validationParameters, callContext);
+            }
+
+            // Key could not be resolved. Depending on the configuration, try all keys or return an error.
+            if (validationParameters.TryAllIssuerSigningKeys)
+                return ValidateSignatureUsingAllKeys(jwtToken, validationParameters, configuration, callContext);
+            else
+                return new SignatureValidationResult(
+                    ValidationFailureType.SignatureValidationFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(TokenLogMessages.IDX10500),
+                        typeof(SecurityTokenSignatureKeyNotFoundException),
+                        new StackFrame()));
+        }
+
+        private static SignatureValidationResult ValidateSignatureUsingAllKeys(
+            JsonWebToken jwtToken,
+            ValidationParameters
+            validationParameters, BaseConfiguration? configuration,
+            CallContext callContext)
+        {
+            // control gets here if:
+            // 1. User specified delegate: IssuerSigningKeyResolver returned null
+            // 2. ResolveIssuerSigningKey returned null
+            // Try all the keys. This is the degenerate case, not concerned about perf.
+            ICollection<SecurityKey> keys = configuration?.SigningKeys ?? new List<SecurityKey>();
+            if (!(keys is List<SecurityKey> keysList))
+                keysList = keys.ToList();
+
+            if (!validationParameters.IssuerSigningKeys.IsNullOrEmpty())
```

Comment:
We don't want to call 'IsNullOrEmpty' as this uses IEnumerable.Any().

ValidationParameters.IssuerSigningKeys is never null and is an IList of string.
You can check for Count and use for( int i =0; ...)